### PR TITLE
gh-103583: Isolate CJK codec modules

### LINKED ISSUE #103583: Isolate _multibytecodec
### Linked PRs
* gh-103540
* gh-103588
* gh-103589
* gh-103868
* gh-103869
* gh-103870

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-103583

### CODE REVIEW

**Comment by erlend-aasland** [comment]
BTW, I removed the `initialized` guards. I'm pretty sure we can do that, but I haven't verified it yet.

**Comment by erlend-aasland** [comment]
Also, we can split out cb2c9e7b43994dd0e8ad1969318fafd1fc0fc040 as a separate PR. That may be beneficial regarding to future bisects; it will also make this PR easier to review.

**Comment by erlend-aasland** [comment]
@vstinner, you want to take a look? :)

**Comment by corona10** [comment]
@erlend-aasland I will take a look by tomorrow

**Comment by erlend-aasland** [comment]
(Pulling in main to get the latest ref. leak fixes by Jelle)

File: Modules/cjkcodecs/cjkcodecs.h [hunk 1]
```diff
@@ -60,11 +60,17 @@ struct pair_encodemap {
...
+#define CJK_MOD_SPECIFIC_STATE
+#endif
+
+typedef struct _cjk_mod_state {
     int num_mappings;
     int num_codecs;
     struct dbcs_map *mapping_list;
     MultibyteCodec *codec_list;
+
+    CJK_MOD_SPECIFIC_STATE
```
Comment (corona10) [inline Modules/cjkcodecs/cjkcodecs.h:73]: ~IIUC, can we always reserve the fields for big5_encmap / big5_decmap ?~
~I am not sure upsides and downsides of both approaches.~

I am trying to understand this code as possible.
I will leave more comments about this later.

Comment (erlend-aasland) [reply Modules/cjkcodecs/cjkcodecs.h:73]: But we also need room for extra module state in the iso2022 codec.

Comment (corona10) [reply Modules/cjkcodecs/cjkcodecs.h:73]: Ah, okay, now I understood the approach.. this is super nice.

Comment (erlend-aasland) [reply Modules/cjkcodecs/cjkcodecs.h:73]: Well, it'a still kind of hacky, but no more hacky than the macro templates are already in place in cjk :)

**Review by corona10 — APPROVED** [review]
LGTM, This is quite a complicated but nice approach. But nothing can be better than this.
Thanks for the hard work!

**Comment by vstinner** [comment]
Well done @erlend-aasland!

**Comment by erlend-aasland** [comment]
> Well done @erlend-aasland!

Thanks, Victor :) This was a tricky one!

**Pull request merged**